diff -Naur linuxcnc.orig/src/emc/usr_intf/axis/scripts/axis.py linuxcnc/src/emc/usr_intf/axis/scripts/axis.py
--- linuxcnc.orig/src/emc/usr_intf/axis/scripts/axis.py	2020-01-17 10:02:18.499019454 +0100
+++ linuxcnc/src/emc/usr_intf/axis/scripts/axis.py	2020-01-17 10:02:19.023024980 +0100
@@ -288,6 +288,9 @@
         self.cache = []
         Tkinter.Frame.__init__(self, master)
 
+    def has_items(self):
+        return (len(self.widgets) > 0) 
+
     def clear(self,iconname=None):
         if iconname:
             cpy = self.widgets[:]
@@ -805,6 +808,7 @@
                      root_window.tk.call("pause_image_override")
                  else:
                      root_window.tk.call("pause_image_normal")
+            comp["has-notifications"] = notifications.has_items()
             for handler in hal_scalehandlers:
                 handler.process()
             if (comp["jog.disable"] or
@@ -3401,6 +3405,7 @@
     comp.newpin("notifications-clear",hal.HAL_BIT,hal.HAL_IN)
     comp.newpin("notifications-clear-info",hal.HAL_BIT,hal.HAL_IN)
     comp.newpin("notifications-clear-error",hal.HAL_BIT,hal.HAL_IN)
+    comp.newpin("has-notifications",hal.HAL_BIT,hal.HAL_OUT)
     comp.newpin("resume-inhibit",hal.HAL_BIT,hal.HAL_IN)
 
     comp.newpin("sliders.scale", hal.HAL_FLOAT, hal.HAL_IN)
