diff -Naur linuxcnc.orig/src/emc/usr_intf/axis/scripts/axis.py linuxcnc/src/emc/usr_intf/axis/scripts/axis.py
--- linuxcnc.orig/src/emc/usr_intf/axis/scripts/axis.py	2020-01-02 20:02:01.000000000 +0100
+++ linuxcnc/src/emc/usr_intf/axis/scripts/axis.py	2020-01-17 10:02:16.907002656 +0100
@@ -2928,8 +2928,10 @@
 postgui_halfile = inifile.find("HAL", "POSTGUI_HALFILE")
 max_feed_override = float(inifile.find("DISPLAY", "MAX_FEED_OVERRIDE"))
 max_spindle_override = float(inifile.find("DISPLAY", "MAX_SPINDLE_OVERRIDE") or max_feed_override)
+min_spindle_override = float(inifile.find("DISPLAY", "MIN_SPINDLE_OVERRIDE") or 0)
 max_feed_override = int(max_feed_override * 100 + 0.5)
 max_spindle_override = int(max_spindle_override * 100 + 0.5)
+min_spindle_override = int(min_spindle_override * 100 + 0.5)
 default_spindle_speed = int(inifile.find("DISPLAY", "DEFAULT_SPINDLE_SPEED") or 1)
 geometry = inifile.find("DISPLAY", "GEOMETRY") or "XYZBCUVW"
 geometry = re.split(" *(-?[XYZABCUVW])", geometry.upper())
@@ -2967,7 +2969,7 @@
 root_window.tk.eval("${pane_top}.maxvel.s set [setval $maxvel_speed $max_maxvel]")
 widgets.feedoverride.configure(to=max_feed_override)
 widgets.rapidoverride.configure(to=100)
-widgets.spinoverride.configure(to=max_spindle_override)
+widgets.spinoverride.configure(from_=min_spindle_override, to=max_spindle_override)
 nmlfile = inifile.find("EMC", "NML_FILE")
 if nmlfile:
     linuxcnc.nmlfile = os.path.join(os.path.dirname(sys.argv[2]), nmlfile)
