diff -Naur linuxcnc.orig/src/emc/rs274ngc/interp_check.cc linuxcnc/src/emc/rs274ngc/interp_check.cc
--- linuxcnc.orig/src/emc/rs274ngc/interp_check.cc	2020-01-02 20:02:01.000000000 +0100
+++ linuxcnc/src/emc/rs274ngc/interp_check.cc	2020-01-17 10:02:26.167100105 +0100
@@ -101,7 +101,12 @@
   } else if (mode1 == G_5_2){
   } else if (mode0 == G_28_1 || mode0 == G_30_1) {
   } else if (mode0 == G_53) {
-    CHKS(((block->motion_to_be != G_0) && (block->motion_to_be != G_1)),
+    CHKS(((block->motion_to_be != G_0) &&
+          (block->motion_to_be != G_1) &&
+          (block->motion_to_be != G_38_2) &&
+          (block->motion_to_be != G_38_3) &&
+          (block->motion_to_be != G_38_4) &&
+          (block->motion_to_be != G_38_5)),
         NCE_MUST_USE_G0_OR_G1_WITH_G53);
     CHKS(((block->g_modes[3] == G_91) ||
          ((block->g_modes[3] != G_90) &&
