diff -Naur linuxcnc.orig/tcl/ngcgui_app.tcl linuxcnc/tcl/ngcgui_app.tcl
--- linuxcnc.orig/tcl/ngcgui_app.tcl	2020-01-02 20:02:01.000000000 +0100
+++ linuxcnc/tcl/ngcgui_app.tcl	2020-01-17 10:02:24.103078438 +0100
@@ -42,6 +42,7 @@
   set program_prefix   [lindex [inifindall DISPLAY  PROGRAM_PREFIX]   end]
   set subroutine_path  [lindex [inifindall RS274NGC SUBROUTINE_PATH]  end]
   set gcmc_include_path [lindex [inifindall DISPLAY GCMC_INCLUDE_PATH]  end]
+  set ngcgui_pentries  [lindex [inifindall DISPLAY NGCGUI_PENTRIES]   end]
 
 
   # allow specification of DISPLAY:NGCGUI for:
@@ -97,7 +98,8 @@
               font=$ngcgui_font \
               startdir=$startdir \
               options=$ngcgui_options \
-              gcmc_include_path=$gcmc_include_path
+              gcmc_include_path=$gcmc_include_path \
+              pentries=$ngcgui_pentries
     incr ::ngcguict
   }
   return
diff -Naur linuxcnc.orig/tcl/ngcgui.tcl linuxcnc/tcl/ngcgui.tcl
--- linuxcnc.orig/tcl/ngcgui.tcl	2020-01-02 20:02:01.000000000 +0100
+++ linuxcnc/tcl/ngcgui.tcl	2020-01-17 10:02:24.103078438 +0100
@@ -983,10 +983,10 @@
 proc ::ngcgui::initgui {hdl} {
   if ![info exists ::ngc(embed,hdl)] {set ::ngc(embed,hdl) 0}
   if [info exists ::ngcgui($hdl,afterid)] { return ;# already done }
+  # number of entries in positional frame 30 max positional parameters
+  # 3 frames max so must have pentries >=10
+  if ![info exists ::ngc(any,pentries)] {set ::ngc(any,pentries) 10}
   # fixed initializations
-  set ::ngc(any,pentries)       10 ;# number of entries in positional frame
-                                   ;# 30 max positional parameters
-                                   ;# 3 frames max so must have pentries >=10
   set ::ngc(any,pollms)         2000
 
   set ::ngc(any,color,black)    black
@@ -3407,6 +3407,7 @@
               font=$::ngc(any,font) \
               options=$::ngc(input,options) \
               gcmc_include_path=$::ngc(input,gcmc_include_path) \
+              pentries=$::ngc(any,pentries) \
              ]
   $::ngc(any,axis,parent) itemconfigure $pageid \
         -createcmd "::ngcgui::pagecreate $newhdl"\
@@ -3645,6 +3646,7 @@
                   startdir \
                   gcmc_include_path \
                   options \
+                  pentries \
                  }
   foreach item $equalitems {set ::ngc(input,$item) ""}
   foreach input  $args {
@@ -3690,6 +3692,7 @@
      set font [string range $font 1 [expr [string len $font] -2]]
   }
   if {"$font" != ""} {set ::ngc(any,font) $font}
+  if {"$pentries" != ""} {set ::ngc(any,pentries) $pentries}
 
   # specific settings for embedding in axis tab:
   set ::ngc(any,aspect)        horiz
